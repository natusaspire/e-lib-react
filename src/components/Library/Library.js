import React, { PureComponent } from 'react';

import {
  MATERIALS_TEXT_ENDPOINT,
  MATERIALS_AUDIO_ENDPOINT,
  MATERIALS_VIDEO_ENDPOINT
} from '@/constants/endpoints';
import { MATERIALS_ORDER } from '@/constants/orderTypes';
import Toolbar from '@/components/Toolbar';
import Tabs from '@/components/Tabs';
import MaterialPreviewList from '@/components/MaterialPreviewList';

const INITIAL_PARAMS = {
  endpoint: MATERIALS_TEXT_ENDPOINT,
  page: 1,
  perPage: 36,
  search: '',
  order: MATERIALS_ORDER.DATE.DESC
};

const PAGE_TITLE = 'Library';

const TABS = [
  {
    id: 1,
    name: 'Text',
    value: MATERIALS_TEXT_ENDPOINT
  },
  {
    id: 2,
    name: 'Audio',
    value: MATERIALS_AUDIO_ENDPOINT
  },
  {
    id: 3,
    name: 'Video',
    value: MATERIALS_VIDEO_ENDPOINT
  }
];

const getCategoryTitle = category => `Library of ${category}`;

class Library extends PureComponent {
  getMaterials = ({
    endpoint = INITIAL_PARAMS.endpoint,
    page = INITIAL_PARAMS.page,
    perPage = INITIAL_PARAMS.perPage,
    search = INITIAL_PARAMS.search,
    order = INITIAL_PARAMS.order
  }) => {
    const { getMaterials } = this.props;

    getMaterials({ endpoint, page, perPage, search, order });
  };

  selectTabItem = ({ value: endpoint }) => this.getMaterials({ endpoint });

  componentDidMount() {
    const { getMaterials } = this.props;

    getMaterials(INITIAL_PARAMS);
  }

  componentDidUpdate() {
    const {
      materials: { endpoint }
    } = this.props;

    document.title = endpoint ? getCategoryTitle(endpoint) : PAGE_TITLE;
  }

  componentWillUnmount() {
    const { clearMaterials } = this.props;

    clearMaterials();
  }

  render() {
    const {
      materials: { endpoint, data }
    } = this.props;

    return (
      <div className="main">
        <div className="main__container main__toolbar--fixed">
          <Toolbar title="Toolbar">
            <Tabs items={TABS} onItemSelect={this.selectTabItem} />
          </Toolbar>
        </div>
        <div className="main__container main__content">
          <MaterialPreviewList endpoint={endpoint} materials={data} />
        </div>
      </div>
    );
  }
}

export default Library;
