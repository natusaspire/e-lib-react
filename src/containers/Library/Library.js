import React, { PureComponent } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { getMaterials, clearMaterials } from '@/actions/materials';
import { MATERIALS_TEXT_ENDPOINT } from '@/constants/endpoints';
import { MATERIALS_ORDER } from '@/constants/orderTypes';
import Library from '@/components/Library';

const INITIAL_PARAMS = {
  endpoint: MATERIALS_TEXT_ENDPOINT,
  page: 1,
  perPage: 36,
  search: '',
  order: MATERIALS_ORDER.DATE.DESC
};

const PAGE_TITLE = 'Library';

const getCategoryTitle = category => `Library of ${category}`;

class LibraryContainer extends PureComponent {
  getMaterials = ({
    endpoint = INITIAL_PARAMS.endpoint,
    page = INITIAL_PARAMS.page,
    perPage = INITIAL_PARAMS.perPage,
    search = INITIAL_PARAMS.search,
    order = INITIAL_PARAMS.order
  }) => {
    const { getMaterials } = this.props;

    getMaterials({ endpoint, page, perPage, search, order });
  };

  componentDidMount() {
    const { getMaterials } = this.props;

    getMaterials(INITIAL_PARAMS);
  }

  componentDidUpdate() {
    const {
      materials: { endpoint }
    } = this.props;

    document.title = endpoint ? getCategoryTitle(endpoint) : PAGE_TITLE;
  }

  componentWillUnmount() {
    const { clearMaterials } = this.props;

    clearMaterials();
  }

  render() {
    const {
      materials: { endpoint, data }
    } = this.props;

    return (
      <Library
        endpoint={endpoint}
        materials={data}
        onUpdate={this.getMaterials}
      />
    );
  }
}

const mapStateToProps = ({ materials }) => ({ materials });

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({ getMaterials, clearMaterials }, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LibraryContainer);
