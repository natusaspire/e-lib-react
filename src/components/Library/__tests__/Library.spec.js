import React from 'react';
import { shallow } from 'enzyme';

import Library from '@/components/Library';

jest.mock('@/constants/endpoints', () => ({
  MATERIALS_TEXT_ENDPOINT: 'text',
  MATERIALS_AUDIO_ENDPOINT: 'audio',
  MATERIALS_VIDEO_ENDPOINT: 'video'
}));

jest.mock('@/components/Toolbar', () => 'Toolbar');
jest.mock('@/components/Tabs', () => 'Tabs');
jest.mock('@/components/MaterialPreviewList', () => 'MaterialPreviewList');

describe('<Library />', () => {
  const props = {
    endpoint: 'text',
    materials: [
      {
        id: 1,
        type: 'text',
        title: 'Material title 1',
        textContent: 'Material text content 1.'
      },
      {
        id: 2,
        type: 'text',
        title: 'Material title 2',
        textContent: 'Material text content 2.'
      },
      {
        id: 3,
        type: 'text',
        title: 'Material title 3',
        textContent: 'Material text content 3.'
      }
    ],
    onUpdate: jest.fn()
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render library content', () => {
    const wrapper = shallow(<Library {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('selectTabItem', () => {
    it('should update endpoint', () => {
      const wrapper = shallow(<Library {...props} />);

      const tab = {
        id: 2,
        name: 'Audio',
        value: 'audio'
      };

      const params = { endpoint: tab.value };

      wrapper.instance().selectTabItem(tab);

      expect(props.onUpdate).toHaveBeenCalledWith(params);
    });
  });
});
