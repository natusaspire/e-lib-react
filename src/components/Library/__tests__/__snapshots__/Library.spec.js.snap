// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Library /> should render library content 1`] = `
<div
  className="main"
>
  <div
    className="main__container main__toolbar--fixed"
  >
    <Toolbar
      title="Toolbar"
    >
      <Tabs
        items={
          Array [
            Object {
              "id": 1,
              "name": "Text",
              "value": "text",
            },
            Object {
              "id": 2,
              "name": "Audio",
              "value": "audio",
            },
            Object {
              "id": 3,
              "name": "Video",
              "value": "video",
            },
          ]
        }
        onItemSelect={[Function]}
      />
    </Toolbar>
  </div>
  <div
    className="main__container main__content"
  >
    <MaterialPreviewList
      endpoint="text"
      materials={
        Array [
          Object {
            "id": 1,
            "textContent": "Material text content 1.",
            "title": "Material title 1",
            "type": "text",
          },
          Object {
            "id": 2,
            "textContent": "Material text content 2.",
            "title": "Material title 2",
            "type": "text",
          },
          Object {
            "id": 3,
            "textContent": "Material text content 3.",
            "title": "Material title 3",
            "type": "text",
          },
        ]
      }
    />
  </div>
</div>
`;
